presto-main/src/main/java/com/facebook/presto/operator/GroupedTopNBuilder.java

```diff
@@ -0,0 +1,654 @@
+        }
+        verify(pageReferences.get(newPageId) == null, "should not overwrite a non-empty slot");
+        pageReferences.set(newPageId, newPageReference);
+
+        // update the affected heaps and record candidate pages that need compaction
+        Set<Integer> pagesToCompact = new IntOpenHashSet();
+        for (int position = 0; position < newPage.getPositionCount(); position++) {
+            long groupId = groupIds != null ? groupIds.getGroupId(position) : 0L;
+            groupedRows.ensureCapacity(groupId + 1);
+
+            RowHeap rows = groupedRows.get(groupId);
+            if (rows == null) {
+                // a new group
+                rows = new RowHeap(Ordering.from(comparator).reversed());
+                groupedRows.set(groupId, rows);
+            }
+            else {
+                // update an existing group;
+                // remove the memory usage for this group for now; add it back after update
+                memorySizeInBytes -= rows.getEstimatedSizeInBytes();
+            }
+
+            if (rows.size() < topN) {
+                // still have space for the current group
+                Row row = new Row(newPageId, position);
+                rows.enqueue(row);
+                newPageReference.reference(row);
+            }
+            else {
+                // may compare with the topN-th element with in the heap to decide if update is necessary
+                Row previousRow = rows.first();
+                Row newRow = new Row(newPageId, position);
+                if (comparator.compare(newRow, previousRow) < 0) {
+                    // update reference and the heap
+                    rows.dequeue();
+                    PageReference previousPageReference = pageReferences.get(previousRow.getPageId());
+                    previousPageReference.dereference(previousRow.getPosition());
+                    newPageReference.reference(newRow);
+                    rows.enqueue(newRow);
+
+                    // compact a page if it is not the current input page and the reference count is below the threshold
+                    if (previousPageReference.getPage() != newPage &&
+                            previousPageReference.getUsedPositionCount() * COMPACT_THRESHOLD < previousPageReference.getPage().getPositionCount() &&
+                            !pagesToCompact.contains(previousRow.getPageId())) {
+                        pagesToCompact.add(previousRow.getPageId());
+                    }
+                }
+            }
+
+            memorySizeInBytes += rows.getEstimatedSizeInBytes();
+        }
+
+        // may compact the new page as well
+        if (newPageReference.getUsedPositionCount() * COMPACT_THRESHOLD < newPage.getPositionCount()) {
+            verify(!pagesToCompact.contains(newPageId));
+            pagesToCompact.add(newPageId);
+        }
+
+        compactPages(pagesToCompact.iterator());
+    }
+
+    private void compactPages(Iterator<Integer> pageIds)
```

Comment:
I'd change this method to take a single `pageId` because I think it makes the callers simpler.  The first caller is making a custom iterator and a simple loop would be easier.  The second caller actually has a primitive `IntIterator` so a custom loop will also will be better.